### PR TITLE
Add validation to check top changelog link

### PR DESCRIPTION
**Pull request by r00tu53r**
## What does this PR do?

On occasion the PR link in the changelog is not updated with the correct number. It is left with `xxxx` or `0` or with no number. The package could get merged to `main` with no check to catch it. This PR adds a validation to check all the changelog entry's links. 

Since the link could technically be any issue tracking system, this change enforces that the base of the path (which is the number) be a valid number (not a `0`) for a github.com URL.

Initial discussion about this is here - https://github.com/elastic/elastic-package/pull/793

## Why is it important?

As explained in the section above.

## Checklist

- [x] I have added unit tests to test the validation.
- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/main/test/packages) that prove my change is effective.
- [x] I have added an entry in [`versions/N/changelog.yml`](https://github.com/elastic/package-spec/blob/main/versions/1/changelog.yml).

### CODE REVIEW

File: versions/1/changelog.yml [hunk 1]
```diff
@@ -2,6 +2,11 @@
 ## This file documents changes in the package specification. It is NOT a package specification file.
 ## Newer entries go at the bottom.
 ##
+- version: 1.7.2-next
```
Comment (mtojek) [inline versions/1/changelog.yml:5]: Let's "park" it in 1.7.1-next. I suppose that all these will hit 1.8.

File: code/go/internal/validator/semantic/version_integrity.go [hunk 2]
```diff
@@ -40,6 +43,20 @@ func ValidateVersionIntegrity(fsys fspath.FS) ve.ValidationErrors {
 	return nil
 }
 
+// ValidateTopChangelogLink returns validation errors if the link is not a valid PR github.com link.
+// If the link is not a github.com link this validation is skipped and does not return an error.
+func ValidateTopChangelogLink(fsys fspath.FS) ve.ValidationErrors {
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity.go:48]: Is there any reason why you validate only the top link? I think it's feasible to go over all entries.

File: code/go/internal/validator/semantic/version_integrity.go [hunk 3]
```diff
@@ -127,3 +154,43 @@ func ensureManifestVersionHasChangelogEntry(manifestVersion string, versions []s
...
+	validateLinks := []struct {
+		domain       string
+		validateLink validateFn
+	}{
+		{
+			"github.com",
+			validateGithub,
+		},
+	}
+	for _, link := range links {
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity.go:171]: This is the reason why we have `ValidationErrors` (plural). You can run the validation against all entries and combine errors into an array.

File: code/go/internal/validator/semantic/version_integrity.go [hunk 4]
```diff
@@ -127,3 +154,43 @@ func ensureManifestVersionHasChangelogEntry(manifestVersion string, versions []s
...
+	return nil
+}
+
+func validateGithub(ghLink *url.URL) error {
+	prNum, err := strconv.Atoi(path.Base(ghLink.Path))
+	if err != nil {
+		return err
+	}
+	if prNum == 0 {
+		return errors.New("zero is not a valid pull request number")
```
Comment (jsoriano) [inline code/go/internal/validator/semantic/version_integrity.go:193]: Nit.
```suggestion
	if prNum >= 0 {
		return errors.New("issue number in changelog link should be a positive number")
```

**Review by jsoriano — COMMENTED** [review]
As mentioned in the original issue, I would prefer not to couple the format link to github, but we can be this strict now and relax later when there is more broad usage.

File: code/go/internal/validator/semantic/version_integrity.go (hunk 2 above)
Comment (r00tu53r) [reply code/go/internal/validator/semantic/version_integrity.go:48]: I wasn't sure if it was okay to validate all the older links as that could break builds for something that was in built a while ago and other changes were built on top of it.

Comment (r00tu53r) [reply code/go/internal/validator/semantic/version_integrity.go]: The code now checks all links.

File: code/go/internal/validator/semantic/version_integrity.go (hunk 3 above)
Comment (r00tu53r) [reply code/go/internal/validator/semantic/version_integrity.go:168]: The individual errors are now being appended to `ValidationErrors`.

File: code/go/internal/validator/semantic/version_integrity.go [hunk 5]
```diff
@@ -127,3 +149,45 @@ func ensureManifestVersionHasChangelogEntry(manifestVersion string, versions []s
 	}
 	return errors.New("current manifest version doesn't have changelog entry")
 }
+
+func ensureLinksAreValid(links []string) ve.ValidationErrors {
+
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity.go:154]: nit: empty line

File: code/go/internal/validator/semantic/version_integrity.go [hunk 6]
```diff
@@ -127,3 +149,45 @@ func ensureManifestVersionHasChangelogEntry(manifestVersion string, versions []s
...
+		}
+		for _, vl := range validateLinks {
+			if strings.Contains(linkURL.Host, vl.domain) {
+				if err = vl.validateLink(linkURL); err != nil {
+					errs.Append(ve.ValidationErrors{err})
+				}
+			}
+		}
+	}
+	return errs
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity.go:184]: I might be wrong, but I thought that we set `errs` when there is at least one error in the array. Otherwise, it's nil.

File: code/go/internal/validator/semantic/version_integrity.go [hunk 7]
```diff
@@ -127,3 +149,45 @@ func ensureManifestVersionHasChangelogEntry(manifestVersion string, versions []s
...
+		if err != nil {
+			errs.Append(ve.ValidationErrors{
+				fmt.Errorf("invalid URL %v", err),
+			})
+			continue
+		}
+		for _, vl := range validateLinks {
+			if strings.Contains(linkURL.Host, vl.domain) {
+				if err = vl.validateLink(linkURL); err != nil {
+					errs.Append(ve.ValidationErrors{err})
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity.go:179]: nit: if `validateFn` returns `ValidationErrors` you wouldn't need the transformation here

File: code/go/internal/validator/semantic/version_integrity_test.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+package semantic
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestChangelogLinks(t *testing.T) {
+
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity_test.go:14]: nit: empty line

File: code/go/internal/validator/semantic/version_integrity_test.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	var tests = []struct {
+		name      string
+		links     []string
+		numErrors int
+	}{
```
Comment (mtojek) [inline code/go/internal/validator/semantic/version_integrity_test.go:19]: Hm... I think you can also prepare table tests with given URLs and expected outcomes (nil or error).

Then, I would add another single unit test to check if `ensureLinksAreValid` iterates over all links (one test with few broken URLs and few valid ones).

File: code/go/internal/validator/semantic/version_integrity.go (hunk 6 above)
Comment (r00tu53r) [reply code/go/internal/validator/semantic/version_integrity.go:184]: Yes for that case the `errs` is nil. For other cases `errs` has errors (if found).

File: code/go/internal/validator/semantic/version_integrity_test.go (hunk 9 above)
Comment (r00tu53r) [reply code/go/internal/validator/semantic/version_integrity_test.go:57]: The code has been refactored to accomodate other link types and error handling.

**Review by mtojek — COMMENTED** [review]
nit-picks

File: code/go/internal/validator/semantic/version_integrity_test.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			errs := ensureLinksAreValid(test.links)
+			if test.errors == nil {
+				assert.Equal(t, test.errors, errs)
+			} else {
+				assert.Equal(t, len(errs), len(test.errors))
+			}
```
[the remark this answers is not in the document]
Comment (mtojek) [reply code/go/internal/validator/semantic/version_integrity_test.go:104]: Ah, ok :) Thanks for checking this.

**Review by mtojek — APPROVED** [review]
LGTM!

**Comment by r00tu53r** [comment]
> LGTM!

@mtojek thanks for the review! I don't seem to have permissions to merge this PR.

**Comment by mtojek** [comment]
Let's wait for review @jsoriano and then we'll merge it.

**Comment by jsoriano** [comment]
Merged, thanks!
